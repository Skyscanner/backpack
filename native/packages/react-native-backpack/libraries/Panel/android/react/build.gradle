apply plugin: 'com.android.library'
apply plugin: 'kotlin-android'

android {
  compileSdkVersion rootProject.ext.compileSdkVersion
  buildToolsVersion rootProject.ext.buildToolsVersion


  defaultConfig {
    minSdkVersion rootProject.ext.minSdkVersion
    targetSdkVersion rootProject.ext.targetSdkVersion
    versionCode 1
    versionName "1.0"
  }
}

dependencies{
  implementation "com.facebook.react:react-native:+" // From node_modules
  implementation project(':panel-native')
  implementation "org.jetbrains.kotlin:kotlin-stdlib-jdk7:$kotlin_version"
}
repositories {
  mavenCentral()
}
