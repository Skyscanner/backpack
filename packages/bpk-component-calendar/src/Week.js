/*
 * Backpack - Skyscanner's Design System
 *
 * Copyright 2017 Skyscanner Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import PropTypes from 'prop-types';
import React, { Component } from 'react';
import { cssModules } from 'bpk-react-utils';

import {
  getDay,
  getFirstDayOfWeekend,
  getLastDayOfWeekend,
  isSameDay,
  isSameWeek,
  isSameMonth,
  isToday,
  isWithinRange,
} from './date-utils';
import CustomPropTypes from './custom-proptypes';
import STYLES from './bpk-calendar-grid.scss';

const getClassName = cssModules(STYLES);

const shallowEqualProps = (props1, props2, propList) => {
  let i = 0;
  for (; i < propList.length; i += 1) {
    if (props1[propList[i]] !== props2[propList[i]]) { return false; }
  }
  return true;
};

/*
  Week - table row containing a week full of DateContainer components
*/
class Week extends Component {
  constructor() {
    super();

    this.onMouseDown = this.onMouseDown.bind(this);
    this.onMouseUp = this.onMouseUp.bind(this);
    this.onMouseClick = this.onMouseClick.bind(this);

    this.state = { savedDate: null };
  }

  componentDidMount() {
    document.addEventListener('mouseup', this.onMouseUp);
  }

  shouldComponentUpdate(nextProps) {
    const shallowProps = [
      'DateComponent',
      'dateModifiers',
      'daysOfWeek',
      'formatDateFull',
      'isKeyboardFocusable',
      'markOutsideDays',
      'markToday',
      'onDateClick',
      'onDateMouseDown',
      'onDateKeyDown',
      'preventKeyboardFocus',
      'showWeekendSeparator',
      'weekStartsOn',
      'dates',
    ];

    if (!shallowEqualProps(this.props, nextProps, shallowProps)) { return true; }
    if (
      (isSameWeek(nextProps.focusedDate, nextProps.dates[0], { weekStartsOn: nextProps.weekStartsOn })
        || isSameWeek(this.props.focusedDate, this.props.dates[0], { weekStartsOn: this.props.weekStartsOn })
      ) && this.props.focusedDate !== nextProps.focusedDate) {
      return true;
    }
    if (
      (isSameWeek(nextProps.selectedDate, nextProps.dates[0], { weekStartsOn: nextProps.weekStartsOn })
        || isSameWeek(this.props.selectedDate, this.props.dates[0], { weekStartsOn: this.props.weekStartsOn })
      ) && this.props.selectedDate !== nextProps.selectedDate) {
      return true;
    }
    if (!isSameDay(nextProps.minDate, this.props.minDate)) {
      return true;
    }
    if (!isSameDay(nextProps.maxDate, this.props.maxDate)) {
      return true;
    }

    return false;
  }

  componentWillUnmount() {
    document.removeEventListener('mouseup', this.onMouseUp);
  }

  onMouseClick(date) {
    this.props.onDateClick(date);
    this.setState({ savedDate: null });
  }

  onMouseDown(date, month) {
    if (isSameMonth(date, month)) {
      this.props.onDateMouseDown(null, { date });
    } else {
      this.setState({ savedDate: date });
    }
  }

  onMouseUp(event) {
    if (this.state.savedDate != null) {
      this.props.onDateMouseDown(event, { date: this.state.savedDate });
      this.setState({ savedDate: null });
    }
  }

  render() {
    const {
      DateComponent,
      dateModifiers,
      daysOfWeek,
      focusedDate,
      formatDateFull,
      isKeyboardFocusable,
      markOutsideDays,
      markToday,
      maxDate,
      minDate,
      month,
      onDateKeyDown,
      preventKeyboardFocus,
      selectedDate,
      showWeekendSeparator,
    } = this.props;

    const firstDayOfWeekendIndex = getFirstDayOfWeekend(daysOfWeek);
    const lastDayOfWeekendIndex = getLastDayOfWeekend(daysOfWeek);

    return (
      <tr
        className={getClassName('bpk-calendar-grid__week')}
      >{ this.props.dates.map(date => (
        <DateContainer
          key={date.getDate()}
          weekendStart={showWeekendSeparator && firstDayOfWeekendIndex === getDay(date)}
          weekendEnd={showWeekendSeparator && lastDayOfWeekendIndex === getDay(date)}
        >
          <DateComponent
            date={date}
            modifiers={dateModifiers}
            aria-label={formatDateFull(date)}
            onClick={() => { this.onMouseClick(date); }}
            onMouseDown={() => { this.onMouseDown(date, month); }}
            onDateKeyDown={onDateKeyDown}
            preventKeyboardFocus={preventKeyboardFocus}
            isKeyboardFocusable={isKeyboardFocusable}
            isFocused={isSameDay(date, focusedDate)}
            isSelected={isSameDay(date, selectedDate)}
            isBlocked={(minDate && maxDate) ? !isWithinRange(date, minDate, maxDate) : false}
            isOutside={markOutsideDays && !isSameMonth(date, month)}
            isToday={markToday && isToday(date)}
          />
        </DateContainer>
      ))}
      </tr>
    );
  }
}

Week.propTypes = {
  DateComponent: PropTypes.func.isRequired,
  dateModifiers: CustomPropTypes.DateModifiers.isRequired,
  dates: PropTypes.arrayOf(Date).isRequired,
  daysOfWeek: CustomPropTypes.DaysOfWeek.isRequired,
  formatDateFull: PropTypes.func.isRequired,
  preventKeyboardFocus: PropTypes.bool.isRequired,
  showWeekendSeparator: PropTypes.bool.isRequired,
  markToday: PropTypes.bool.isRequired,
  markOutsideDays: PropTypes.bool.isRequired,
  isKeyboardFocusable: PropTypes.bool.isRequired,
  month: PropTypes.instanceOf(Date).isRequired,
  weekStartsOn: PropTypes.number.isRequired,
  focusedDate: PropTypes.instanceOf(Date),
  maxDate: PropTypes.instanceOf(Date),
  minDate: PropTypes.instanceOf(Date),
  onDateClick: PropTypes.func,
  onDateMouseDown: PropTypes.func,
  onDateKeyDown: PropTypes.func,
  selectedDate: PropTypes.instanceOf(Date),
};

Week.defaultProps = {
  focusedDate: null,
  maxDate: null,
  minDate: null,
  onDateClick: null,
  onDateMouseDown: null,
  onDateKeyDown: null,
  selectedDate: null,
};

/*
  DateContainer - one for each date in the grid; wraps the actual BpkCalendarDate (or custom) component
*/
const DateContainer = (props) => {
  const classNames = [getClassName('bpk-calendar-grid__date')];

  if (props.weekendStart) { classNames.push(getClassName('bpk-calendar-grid__date--weekend-start')); }
  if (props.weekendEnd) { classNames.push(getClassName('bpk-calendar-grid__date--weekend-end')); }

  return (
    <td
      className={classNames.join(' ')}
    >
      { props.children }
    </td>
  );
};

DateContainer.propTypes = {
  children: PropTypes.element.isRequired,
  weekendStart: PropTypes.bool.isRequired,
  weekendEnd: PropTypes.bool.isRequired,
};

export default Week;
