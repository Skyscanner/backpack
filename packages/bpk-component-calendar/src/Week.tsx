/*
 * Backpack - Skyscanner's Design System
 *
 * Copyright 2016 Skyscanner Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import type { ElementType, ReactElement } from 'react';
import { Component } from 'react';

import { cssModules } from '../../bpk-react-utils';

import { SELECTION_TYPES } from './BpkCalendarDate';
import { CALENDAR_SELECTION_TYPE } from './custom-proptypes';
import {
  isSameDay,
  isSameWeek,
  isSameMonth,
  isToday,
  isWithinRange,
  startOfMonth,
  endOfMonth,
} from './date-utils';

import type {
  DateModifiers,
  SelectionConfiguration,
  SelectionConfigurationSingle,
  SelectionConfigurationRange,
} from './custom-proptypes';

// TODO: Move this to `Week.scss`
// This should be using its own css file as `BpkCalendarGrid` is also importing `BpkCalendarGrid.module.scss`
// and the order of css imports can break the component.
import STYLES from './BpkCalendarGrid.module.scss';

const getClassName = cssModules(STYLES);

const shallowEqualProps = (
  props1: Props,
  props2: Props,
  propList: string[],
) => {
  let i = 0;
  for (; i < propList.length; i += 1) {
    if (props1[propList[i]] !== props2[propList[i]]) {
      return false;
    }
  }
  return true;
};

function or(total: boolean, bool: boolean) {
  return total || bool;
}

/**
 * Determines if the current date is selected
 * @param {Date} date the current date from the calendar
 * @param {Object} selectionConfiguration the current selection configuration
 * @returns {Boolean} true is selected and false if not
 */
function getSelectedDate(
  date: Date,
  selectionConfiguration: SelectionConfiguration,
) {
  let startDate;
  let endDate;
  if (selectionConfiguration.type === CALENDAR_SELECTION_TYPE.single) {
    startDate = selectionConfiguration.date;
  } else {
    startDate = selectionConfiguration.startDate;
    endDate = selectionConfiguration.endDate;
  }
  const sameStartDay = startDate && isSameDay(date, startDate);
  const sameEndDay = endDate && isSameDay(date, endDate);

  if (
    (startDate &&
      endDate &&
      (sameStartDay ||
        sameEndDay ||
        isWithinRange(date, { start: startDate, end: endDate }))) ||
    (startDate && !endDate && sameStartDay) ||
    sameStartDay
  ) {
    return true;
  }

  return false;
}

/**
 * Gets the correct selection type for the current date
 * @param {Date} date the current date of the calendar
 * @param {Object} selectionConfiguration the current selection configuration
 * @param {Function} formatDateFull function to format dates
 * @param {Date} month the current month of the calendar
 * @param {Number} weekStartsOn index of the first day of the week
 * @param {Boolean} ignoreOutsideDate ignore date outside current month
 * @returns {String} selection type to be passed to the date
 */
function getSelectionType(
  date: Date,
  selectionConfiguration: SelectionConfiguration,
  formatDateFull: (d: Date) => Date | string,
  month: Date,
  weekStartsOn: 0 | 1 | 2 | 3 | 4 | 5 | 6,
  ignoreOutsideDate: boolean,
) {
  const { endDate, startDate } =
    selectionConfiguration.type === CALENDAR_SELECTION_TYPE.range
      ? selectionConfiguration
      : { startDate: null, endDate: null };

  const sameStartDay = startDate && isSameDay(date, startDate);
  const sameEndDay = endDate && isSameDay(date, endDate);
  const rangeDates = startDate && endDate;
  const isEmptyCell = !isSameMonth(date, month) && ignoreOutsideDate;

  if (
    selectionConfiguration.type === CALENDAR_SELECTION_TYPE.single &&
    selectionConfiguration.date &&
    (selectionConfiguration.date === formatDateFull(date) ||
      formatDateFull(selectionConfiguration.date) === formatDateFull(date))
  ) {
    return SELECTION_TYPES.single;
  }
  if (selectionConfiguration.type === CALENDAR_SELECTION_TYPE.range) {
    if (
      (startDate && !endDate && sameStartDay) ||
      (rangeDates && sameStartDay && sameEndDay)
    ) {
      return SELECTION_TYPES.single;
    }
    if (
      isEmptyCell &&
      rangeDates &&
      ((isSameWeek(startDate, startOfMonth(startDate), { weekStartsOn }) &&
        isSameWeek(date, startDate, { weekStartsOn })) ||
        (isSameWeek(endDate, endOfMonth(endDate), { weekStartsOn }) &&
          isSameWeek(date, endDate, { weekStartsOn })))
    ) {
      return SELECTION_TYPES.none;
    }
    if (
      isEmptyCell &&
      rangeDates &&
      !isSameMonth(startDate, endDate) &&
      ((isSameWeek(startDate, endOfMonth(startDate), { weekStartsOn }) &&
        isSameWeek(date, startDate, { weekStartsOn })) ||
        (isSameWeek(endDate, startOfMonth(endDate), { weekStartsOn }) &&
          isSameWeek(date, endDate, { weekStartsOn })))
    ) {
      return SELECTION_TYPES.middle;
    }
    if (
      rangeDates &&
      isWithinRange(date, { start: startDate, end: endDate }) &&
      !sameStartDay &&
      !sameEndDay
    ) {
      return SELECTION_TYPES.middle;
    }
    if (startDate && formatDateFull(startDate) === formatDateFull(date)) {
      return SELECTION_TYPES.start;
    }
    if (endDate && formatDateFull(endDate) === formatDateFull(date)) {
      return SELECTION_TYPES.end;
    }
  }

  return SELECTION_TYPES.none;
}

/**
 * Handles changing selected date when in single mode
 * @param {Object} props current props
 * @param {Object} nextProps next props as the component is updating
 * @returns {Boolean} based on if the date has changed
 */
const singleDateHandler = (props: Props, nextProps: Props) => {
  const currentSelectConfig =
    props.selectionConfiguration as SelectionConfigurationSingle;
  const nextSelectConfig =
    nextProps.selectionConfiguration as SelectionConfigurationSingle;

  if (
    ((nextSelectConfig.date &&
      isSameWeek(nextSelectConfig.date, nextProps.dates[0], {
        weekStartsOn: nextProps.weekStartsOn,
      })) ||
      (currentSelectConfig.date &&
        isSameWeek(currentSelectConfig.date, props.dates[0], {
          weekStartsOn: props.weekStartsOn,
        }))) &&
    currentSelectConfig.date !== nextSelectConfig.date
  ) {
    return true;
  }
  return false;
};

/**
 * Handles changing selected date when in range mode
 * @param {Object} props current props
 * @param {Object} nextProps next props as the component is updating
 * @returns {Boolean} based on if the date has changed
 */
const rangeDateHandler = (props: Props, nextProps: Props) => {
  const { endDate, startDate } =
    props.selectionConfiguration as SelectionConfigurationRange;
  const { endDate: nextEndDate, startDate: nextStartDate } =
    nextProps.selectionConfiguration as SelectionConfigurationRange;

  const startDateChanged =
    startDate && nextStartDate && !isSameDay(startDate, nextStartDate);
  const endDateChanged =
    endDate && nextEndDate && !isSameDay(endDate, nextEndDate);

  if (startDateChanged || endDateChanged) {
    return true;
  }

  return false;
};

export type Props = DefaultProps & {
  DateComponent: ElementType;
  dateModifiers: DateModifiers;
  dates: Date[];
  formatDateFull: (date: Date) => Date | string;
  preventKeyboardFocus: boolean;
  markToday: boolean;
  markOutsideDays: boolean;
  isKeyboardFocusable: boolean;
  month: Date;
  weekStartsOn: 0 | 1 | 2 | 3 | 4 | 5 | 6;
  [rest: string]: any; // Inexact rest. See decisions/inexact-rest.md
};

type DefaultProps = {
  dateProps?: {};
  focusedDate?: Date | number | null;
  ignoreOutsideDate?: boolean;
  maxDate?: Date | null;
  minDate?: Date | null;
  onDateClick?: () => void;
  onDateKeyDown?: () => void;
  cellClassName?: string | null;
  selectionConfiguration?: SelectionConfiguration;
};

/*
  Week - table row containing a week full of DateContainer components
*/
class Week extends Component<Props> {
  static defaultProps: DefaultProps = {
    dateProps: {},
    focusedDate: null,
    ignoreOutsideDate: false,
    maxDate: null,
    minDate: null,
    onDateClick: () => {},
    onDateKeyDown: () => {},
    selectionConfiguration: {
      type: CALENDAR_SELECTION_TYPE.single,
      date: null,
    },
    cellClassName: null,
  };

  shouldComponentUpdate(nextProps: Props) {
    const shallowProps = [
      'DateComponent',
      'dateModifiers',
      'formatDateFull',
      'isKeyboardFocusable',
      'markOutsideDays',
      'markToday',
      'onDateClick',
      'onDateKeyDown',
      'preventKeyboardFocus',
      'weekStartsOn',
      'dates',
      'cellClassName',
    ];

    // If any of the props have changed, component should update.
    if (!shallowEqualProps(this.props, nextProps, shallowProps)) {
      return true;
    }

    // If focusedDate is changing, and it'll be included as part
    // of either the week we're rendering now or the next week
    // we'll render, component should update.
    if (
      ((nextProps.focusedDate &&
        isSameWeek(nextProps.focusedDate, nextProps.dates[0], {
          weekStartsOn: nextProps.weekStartsOn,
        })) ||
        (this.props.focusedDate &&
          isSameWeek(this.props.focusedDate, this.props.dates[0], {
            weekStartsOn: this.props.weekStartsOn,
          }))) &&
      this.props.focusedDate !== nextProps.focusedDate
    ) {
      return true;
    }

    // If selected date is changing, and it'll be included as part
    // of either the week we're rendering now or the next week we'll
    // render, component should update.
    if (
      this.props.selectionConfiguration?.type ===
        CALENDAR_SELECTION_TYPE.single &&
      this.props.selectionConfiguration.date
    ) {
      return singleDateHandler(this.props, nextProps);
    }
    if (
      this.props.selectionConfiguration?.type === CALENDAR_SELECTION_TYPE.range
    ) {
      return rangeDateHandler(this.props, nextProps);
    }

    // If min date is changing, component should update.
    if (nextProps.minDate) {
      if (nextProps.minDate !== this.props.minDate) {
        return true;
      }
      if (!isSameDay(nextProps.minDate, this.props.minDate)) {
        return true;
      }
    }

    // If max date is changing, component should update.
    if (nextProps.maxDate) {
      if (nextProps.maxDate !== this.props.maxDate) {
        return true;
      }
      if (!isSameDay(nextProps.maxDate, this.props.maxDate)) {
        return true;
      }
    }

    return false;
  }

  render() {
    const {
      DateComponent,
      cellClassName,
      dateModifiers,
      dateProps,
      focusedDate,
      formatDateFull,
      ignoreOutsideDate,
      isKeyboardFocusable,
      markOutsideDays,
      markToday,
      maxDate,
      minDate,
      month,
      onDateClick,
      onDateKeyDown,
      preventKeyboardFocus,
      selectionConfiguration,
      weekStartsOn,
    } = this.props;

    if (ignoreOutsideDate) {
      const daysOutside = this.props.dates.map((date) =>
        isSameMonth(date, month),
      );

      const shouldRender = daysOutside.reduce(or);

      if (!shouldRender) {
        return null;
      }
    }

    return (
      <div className={getClassName('bpk-calendar-grid__week')}>
        {this.props.dates.map((date) => {
          const isBlocked =
            minDate && maxDate
              ? !isWithinRange(date, { start: minDate, end: maxDate })
              : false;

          const dateSelectionType = getSelectionType(
            date,
            selectionConfiguration!,
            formatDateFull,
            month,
            weekStartsOn,
            ignoreOutsideDate!,
          );

          return (
            <DateContainer
              className={cellClassName}
              isEmptyCell={!isSameMonth(date, month) && ignoreOutsideDate!}
              isBlocked={isBlocked}
              key={date.getDate()}
              selectionType={dateSelectionType}
            >
              <DateComponent
                date={date}
                modifiers={dateModifiers}
                aria-label={formatDateFull(date)}
                onClick={onDateClick}
                onDateKeyDown={onDateKeyDown}
                preventKeyboardFocus={preventKeyboardFocus}
                isKeyboardFocusable={isKeyboardFocusable}
                isFocused={focusedDate && isSameDay(date, focusedDate)}
                isSelected={getSelectedDate(date, selectionConfiguration!)}
                isBlocked={isBlocked}
                isOutside={markOutsideDays && !isSameMonth(date, month)}
                isToday={markToday && isToday(date)}
                selectionType={dateSelectionType}
                {...dateProps}
              />
            </DateContainer>
          );
        })}
      </div>
    );
  }
}

type DateContainerProps = {
  children: ReactElement;
  className?: string | null;
  isBlocked: boolean;
  isEmptyCell: boolean;
  selectionType: string;
};
/*
  DateContainer - one for each date in the grid; wraps the actual BpkCalendarDate (or custom) component
*/
const DateContainer = ({
  children,
  className = null,
  isBlocked,
  isEmptyCell,
  selectionType,
}: DateContainerProps) => {
  const classNames = getClassName(
    'bpk-calendar-grid__date',
    `bpk-calendar-grid__date--${selectionType}`,
    className,
  );

  return (
    <div aria-hidden={isEmptyCell || isBlocked} className={classNames}>
      {children}
    </div>
  );
};

export default Week;
