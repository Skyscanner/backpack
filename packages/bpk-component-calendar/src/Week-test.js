/*
 * Backpack - Skyscanner's Design System
 *
 * Copyright 2016 Skyscanner Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import { render } from '@testing-library/react';
import '@testing-library/jest-dom';
import startOfDay from 'date-fns/startOfDay';

import { CALENDAR_SELECTION_TYPE, SELECTION_TYPES } from '..';

import Week from './Week';

const mockDateComponent = jest.fn();

const DummyDateComponent = (props) => {
  mockDateComponent(props);
  return <div />;
};

const initialProps = {
  ...Week.defaultProps,
  DateComponent: DummyDateComponent,
  dateModifiers: {},
  dates: [
    new Date(1980, 4, 31),
    new Date(1980, 5, 1),
    new Date(1980, 5, 2),
    new Date(1980, 5, 3),
    new Date(1980, 5, 4),
    new Date(1980, 5, 5),
    new Date(1980, 5, 6),
  ].map(startOfDay),
  formatDateFull: (d) => d.toString(),
  preventKeyboardFocus: false,
  markToday: true,
  markOutsideDays: true,
  isKeyboardFocusable: true,
  month: new Date(1980, 5, 1),
  weekStartsOn: 0,
};

describe('Week', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should render correctly', () => {
    const { asFragment } = render(<Week {...initialProps} />);

    expect(asFragment()).toMatchSnapshot();
  });

  it('should set "isOutside" prop correctly when "markOutsideDays" attribute equals to "true"', () => {
    render(<Week {...initialProps} markOutsideDays />);

    expect(mockDateComponent).toHaveBeenCalledWith(
      expect.objectContaining({
        date: new Date(1980, 4, 31),
        isOutside: true,
      }),
    );
  });

  it('should set "isSelected" prop correctly when selectionConfiguration type is "range"', () => {
    render(
      <Week
        {...initialProps}
        selectionConfiguration={{
          type: CALENDAR_SELECTION_TYPE.range,
          startDate: new Date(1980, 5, 1),
          endDate: new Date(1980, 5, 3),
        }}
      />,
    );

    expect(mockDateComponent).toHaveBeenCalledWith(
      expect.objectContaining({
        date: new Date(1980, 5, 1),
        isSelected: true,
      }),
    );
    expect(mockDateComponent).toHaveBeenCalledWith(
      expect.objectContaining({
        date: new Date(1980, 5, 2),
        isSelected: true,
      }),
    );
    expect(mockDateComponent).toHaveBeenCalledWith(
      expect.objectContaining({
        date: new Date(1980, 5, 3),
        isSelected: true,
      }),
    );
  });

  it('should set "isSelected" prop correctly when selectionConfiguration type is "single"', () => {
    render(
      <Week
        {...initialProps}
        selectionConfiguration={{
          type: CALENDAR_SELECTION_TYPE.single,
          date: new Date(1980, 5, 5),
        }}
      />,
    );

    expect(mockDateComponent).toHaveBeenCalledWith(
      expect.objectContaining({
        date: new Date(1980, 5, 5),
        isSelected: true,
      }),
    );
  });

  it('should not have any selected dates when selectionConfiguration has null date', () => {
    render(
      <Week
        {...initialProps}
        selectionConfiguration={{
          type: CALENDAR_SELECTION_TYPE.single,
          date: null,
        }}
      />,
    );

    expect(mockDateComponent).not.toHaveBeenCalledWith(
      expect.objectContaining({
        isSelected: true,
      }),
    );
  });

  it('should set "selectionType" prop correctly when selectionConfiguration type is "single"', () => {
    render(
      <Week
        {...initialProps}
        selectionConfiguration={{
          type: CALENDAR_SELECTION_TYPE.single,
          date: new Date(1980, 5, 5),
        }}
      />,
    );

    expect(mockDateComponent).toHaveBeenCalledWith(
      expect.objectContaining({
        date: new Date(1980, 5, 5),
        selectionType: SELECTION_TYPES.single,
      }),
    );
  });

  it('should set "selectionType" prop correctly when selectionConfiguration type is "range" and has no "endDate"', () => {
    render(
      <Week
        {...initialProps}
        selectionConfiguration={{
          type: CALENDAR_SELECTION_TYPE.range,
          startDate: new Date(1980, 5, 5),
          endDate: null,
        }}
      />,
    );

    expect(mockDateComponent).toHaveBeenCalledWith(
      expect.objectContaining({
        date: new Date(1980, 5, 5),
        selectionType: SELECTION_TYPES.single,
      }),
    );
  });

  it('should set "selectionType" prop correctly when selectionConfiguration type is "range"', () => {
    render(
      <Week
        {...initialProps}
        selectionConfiguration={{
          type: CALENDAR_SELECTION_TYPE.range,
          startDate: new Date(1980, 5, 4),
          endDate: new Date(1980, 5, 6),
        }}
      />,
    );

    expect(mockDateComponent).toHaveBeenCalledWith(
      expect.objectContaining({
        date: new Date(1980, 5, 4),
        selectionType: SELECTION_TYPES.start,
      }),
    );

    expect(mockDateComponent).toHaveBeenCalledWith(
      expect.objectContaining({
        date: new Date(1980, 5, 5),
        selectionType: SELECTION_TYPES.middle,
      }),
    );

    expect(mockDateComponent).toHaveBeenCalledWith(
      expect.objectContaining({
        date: new Date(1980, 5, 6),
        selectionType: SELECTION_TYPES.end,
      }),
    );
  });
});
