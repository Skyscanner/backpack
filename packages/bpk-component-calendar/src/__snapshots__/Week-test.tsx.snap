// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Week should render correctly 1`] = `
<DocumentFragment>
  <div
    class="bpk-calendar-grid__week"
  >
    <div
      aria-hidden="false"
      class="bpk-calendar-grid__date bpk-calendar-grid__date--none"
    >
      <div />
    </div>
    <div
      aria-hidden="false"
      class="bpk-calendar-grid__date bpk-calendar-grid__date--none"
    >
      <div />
    </div>
    <div
      aria-hidden="false"
      class="bpk-calendar-grid__date bpk-calendar-grid__date--none"
    >
      <div />
    </div>
    <div
      aria-hidden="false"
      class="bpk-calendar-grid__date bpk-calendar-grid__date--none"
    >
      <div />
    </div>
    <div
      aria-hidden="false"
      class="bpk-calendar-grid__date bpk-calendar-grid__date--none"
    >
      <div />
    </div>
    <div
      aria-hidden="false"
      class="bpk-calendar-grid__date bpk-calendar-grid__date--none"
    >
      <div />
    </div>
    <div
      aria-hidden="false"
      class="bpk-calendar-grid__date bpk-calendar-grid__date--none"
    >
      <div />
    </div>
  </div>
</DocumentFragment>
`;
